File: app/services/tax.ts
import { apiCall } from '@/lib/api';

export interface TaxReportDetails {
  id: number;
  ticker: string;
  date: string;
  quantity: number;
  sale_value: number;
  cost_value: number;
  profit: number;
  is_fii: boolean;
  is_day_trade: boolean;
}

export interface TaxMonthlyReport {
  period: string;
  sales_total_stocks: number;
  is_exempt_stocks_st: boolean;
  profits: {
    stocks_st: number;
    stocks_dt: number;
    fiis: number;
  };
  taxable_profits: {
    stocks_st: number;
    stocks_dt: number;
    fiis: number;
  };
  taxes: {
    stocks_st: number;
    stocks_dt: number;
    fiis: number;
  };
  darf: {
    month_tax: number;
    previous_accumulated: number;
    total_due: number;
    darf_to_pay: number;
    next_month_accumulated: number;
  };
  accumulated_losses: {
    stocks_st: number;
    stocks_dt: number;
    fiis: number;
  };
  details: TaxReportDetails[];
}

export interface AnnualTaxReport {
  year: number;
  bens_e_direitos: {
    ticker: string;
    name: string;
    cnpj: string;
    category: string;
    quantity: number;
    total_cost: number;
    average_price: number;
    description: string;
  }[];
  rendimentos_isentos: {
    dividendos: number;
    lucro_vendas_20k: number;
    total: number;
  };
  rendimentos_exclusivos: {
    jcp: number;
    total: number;
  };
  renda_variavel: {
    month: number;
    profit_st: number;
    profit_dt: number;
    profit_fii: number;
    tax_due: number;
    irrf_st: number;
    irrf_dt: number;
    irrf_fii: number;
    is_exempt_st: boolean;
  }[];
}

export interface TaxProfile {
  accumulated_loss_stocks_st: number;
  accumulated_loss_stocks_dt: number;
  accumulated_loss_fiis: number;
  accumulated_darf_balance: number;
}

export const taxService = {
  getMonthlyTax: async (year: number, month: number): Promise<TaxMonthlyReport> => {
    return await apiCall<TaxMonthlyReport>(`/api/tax/monthly?year=${year}&month=${month}`);
  },

  getAnnualTax: async (year: number): Promise<AnnualTaxReport> => {
    return await apiCall<AnnualTaxReport>(`/api/tax/annual?year=${year}`);
  },

  getTaxProfile: async (): Promise<TaxProfile> => {
    return await apiCall<TaxProfile>('/api/tax/profile');
  },

  updateTaxProfile: async (data: Partial<TaxProfile>): Promise<{message: string}> => {
    return await apiCall<{message: string}>('/api/tax/profile', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }
};

File: server/routes/tax.py
from flask import Blueprint, jsonify, request, g
from db.session import Session
from domain.tax.tax_engine import calculate_monthly_darf, get_or_create_tax_profile

tax_bp = Blueprint('tax', __name__)

@tax_bp.route('/api/tax/monthly', methods=['GET'])
def get_monthly_tax():
    year = request.args.get('year', type=int)
    month = request.args.get('month', type=int)
    
    if not year or not month:
        return jsonify({"error": "year and month are required"}), 400
        
    try:
        user_id = getattr(g, 'user_id', 1)
        report = calculate_monthly_darf(user_id, month, year)
        return jsonify(report), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

@tax_bp.route('/api/tax/profile', methods=['GET'])
def get_tax_profile():
    user_id = getattr(g, 'user_id', 1)
    with Session() as session:
        profile = get_or_create_tax_profile(session, user_id)
        return jsonify({
            "accumulated_loss_stocks_st": float(profile.accumulated_loss_stocks_st),
            "accumulated_loss_stocks_dt": float(profile.accumulated_loss_stocks_dt),
            "accumulated_loss_fiis": float(profile.accumulated_loss_fiis),
            "accumulated_darf_balance": float(profile.accumulated_darf_balance)
        }), 200

@tax_bp.route('/api/tax/profile', methods=['POST'])
def update_tax_profile():
    user_id = getattr(g, 'user_id', 1)
    data = request.json or {}
    with Session() as session:
        profile = get_or_create_tax_profile(session, user_id)
        
        if 'accumulated_loss_stocks_st' in data:
            profile.accumulated_loss_stocks_st = data['accumulated_loss_stocks_st']
        if 'accumulated_loss_stocks_dt' in data:
            profile.accumulated_loss_stocks_dt = data['accumulated_loss_stocks_dt']
        if 'accumulated_loss_fiis' in data:
            profile.accumulated_loss_fiis = data['accumulated_loss_fiis']
        if 'accumulated_darf_balance' in data:
            profile.accumulated_darf_balance = data['accumulated_darf_balance']
            
        session.commit()
        return jsonify({"message": "Tax profile updated successfully"}), 200

@tax_bp.route('/api/tax/annual', methods=['GET'])
def get_annual_tax():
    from domain.tax.annual_irpf import calculate_annual_irpf
    year = request.args.get('year', type=int)
    
    if not year:
        return jsonify({"error": "year is required"}), 400
        
    try:
        user_id = getattr(g, 'user_id', 1)
        report = calculate_annual_irpf(user_id, year)
        return jsonify(report), 200
    except Exception as e:
        import traceback
        traceback.print_exc()
        return jsonify({"error": str(e)}), 500
